File: frameworks/typescript-playwright/services/order_service.ts
import { expect, type APIRequestContext } from '@playwright/test';
import { environment } from '../config/environment.js';
import { teachingData } from '../data/test_data.js';
import { uniqueProductName } from '../helpers/unique_name.js';
import { ProductService, type Product } from './product_service.js';

export type Order = {
  id: number;
  subtotal: number;
  tax: number;
  shipping: number;
  total: number;
  status: string;
  payment_status: string;
  order_items?: Array<Record<string, unknown>>;
  order_shipping?: Record<string, unknown> | Array<Record<string, unknown>>;
};

export type ControlledOrder = { order: Order; product: Product };

export class OrderService {
  constructor(
    private readonly request: APIRequestContext,
    private readonly products: ProductService,
  ) {}

  private headers() {
    return { Authorization: `Bearer ${environment.apiKey}` };
  }

  async createControlledOrder(tag: string, paymentStatus = 'paid'): Promise<ControlledOrder> {
    await this.products.clearCart();
    const product = await this.products.createProduct({
      name: uniqueProductName(tag),
      ...teachingData.product,
    });
    const add = await this.request.post('/api/cart', {
      headers: this.headers(),
      data: { product_id: product.id, quantity: 1 },
    });
    expect(add.status()).toBe(201);

    const shipping = {
      full_name: 'Framework Test User',
      email: `framework.${Date.now()}@example.test`,
      address_line: '123 Main Street',
      city: 'New York',
      state: 'NY',
      zip_code: '10001',
      country: 'USA',
      is_fake: true,
    };
    const payment = {
      card_number: '4242424242424242',
      expiry_date: '12/25',
      cvv: '123',
      brand: 'Visa',
      payment_status: paymentStatus,
    };
    let financial = { tax: 0, shipping: 0, state: 'NY', country: 'USA' };
    let response = await this.request.post('/api/orders', {
      headers: this.headers(),
      data: { shipping, payment, financial },
    });
    if (response.status() === 400) {
      const mismatch = await response.json();
      expect(mismatch.error).toBe('SYNC_ERROR');
      financial = {
        tax: mismatch.correct_values.tax,
        shipping: mismatch.correct_values.shipping_cost,
        state: 'NY',
        country: 'USA',
      };
      response = await this.request.post('/api/orders', {
        headers: this.headers(),
        data: { shipping, payment, financial },
      });
    }
    expect(response.status(), await response.text()).toBe(201);
    const created = (await response.json()).data as Order & { order_id: number };
    const order = { ...created, id: created.order_id };
    return { order, product };
  }

  async get(id: number): Promise<Order> {
    const response = await this.request.get(`/api/orders/${id}`, { headers: this.headers() });
    expect(response.status()).toBe(200);
    return (await response.json()).data as Order;
  }

  async cleanup(controlled?: ControlledOrder, orderAlreadyDeleted = false): Promise<void> {
    if (!controlled) return;
    if (!orderAlreadyDeleted) await this.products.deleteOrder(controlled.order.id);
    await this.products.clearCart();
    await this.products.deleteProduct(controlled.product.id);
  }
}

File: frameworks/java-selenium-rest-assured/services/OrderService.java
package services;

import static io.restassured.RestAssured.given;

import config.Environment;
import data.TestData;
import helpers.UniqueName;
import io.restassured.http.ContentType;
import io.restassured.response.Response;
import java.util.LinkedHashMap;
import java.util.Map;

/** Creates an isolated persisted order for tests that must not depend on shared history. */
public final class OrderService {
    public record ControlledOrder(Map<String, Object> order, Map<String, Object> product) {
        public int orderId() { return ((Number) order.get("id")).intValue(); }
        public int productId() { return ((Number) product.get("id")).intValue(); }
    }

    private final ProductService products = new ProductService();

    private Map<String, String> headers() {
        Map<String, String> headers = new LinkedHashMap<>(Environment.automationHeaders());
        headers.put("Authorization", "Bearer " + Environment.required("API_KEY"));
        return headers;
    }

    public ControlledOrder createControlledOrder() {
        products.clearCart();
        Map<String, Object> product = products.createProduct(
                ProductService.newProduct(UniqueName.product(), TestData.PRODUCT));
        try {
            Response added = products.addToCart(((Number) product.get("id")).intValue(), 1);
            if (added.statusCode() != 201) throw new AssertionError(added.asString());
            Map<String, Object> shipping = Map.of(
                    "full_name", "Framework Test User", "email", "framework." + System.nanoTime() + "@example.test",
                    "address_line", "123 Main Street", "city", "New York", "state", "NY",
                    "zip_code", "10001", "country", "USA", "is_fake", true);
            Map<String, Object> payment = Map.of("card_number", "4242424242424242",
                    "expiry_date", "12/25", "cvv", "123", "brand", "Visa", "payment_status", "paid");
            Map<String, Object> financial = Map.of("tax", 0, "shipping", 0, "state", "NY", "country", "USA");
            Response response = create(shipping, payment, financial);
            if (response.statusCode() == 400 && "SYNC_ERROR".equals(response.jsonPath().getString("error"))) {
                financial = Map.of("tax", response.jsonPath().getDouble("correct_values.tax"),
                        "shipping", response.jsonPath().getDouble("correct_values.shipping_cost"),
                        "state", "NY", "country", "USA");
                response = create(shipping, payment, financial);
            }
            if (response.statusCode() != 201) throw new AssertionError(response.asString());
            Map<String, Object> order = new LinkedHashMap<>(response.jsonPath().getMap("data"));
            order.put("id", order.get("order_id"));
            return new ControlledOrder(order, product);
        } catch (RuntimeException | AssertionError error) {
            products.clearCart(); products.deleteProduct(((Number) product.get("id")).intValue());
            throw error;
        }
    }

    private Response create(Map<String, Object> shipping, Map<String, Object> payment,
            Map<String, Object> financial) {
        return given().headers(headers()).contentType(ContentType.JSON)
                .body(Map.of("shipping", shipping, "payment", payment, "financial", financial))
                .post(Environment.BASE_URL + "/api/orders");
    }

    public Map<String, Object> get(int id) {
        Response response = given().headers(headers()).get(Environment.BASE_URL + "/api/orders/" + id);
        if (response.statusCode() != 200) throw new AssertionError(response.asString());
        return response.jsonPath().getMap("data");
    }

    public void cleanup(ControlledOrder controlled) {
        if (controlled == null) return;
        products.deleteOrder(controlled.orderId()); products.clearCart(); products.deleteProduct(controlled.productId());
    }
}
